# King County House Sales Prediction Project


**Authors**: [Dominic Muli](mailto:dominic.muli@student.moringaschool.com), [Allan Ngeiywa](mailto:allan.ngeiywa@student.moringaschool.com), [Celestine Imelda](mailto:celestine.imelda@student.moringaschool.com), and [Kevin Kagia](mailto:kevin.kagia@student.moringaschool.com)


## 1.0) Project Overview


<img src="./images/kingcountylogo_mlk.webp" alt="KingCounty" height="100">

[King County](https://kingcounty.gov/en) situated in the U.S. state of Washington boasts a vibrant real estate market. This project conducts a comprehensive exploration of King County's housing landscape, aiming to unravel the factors that influence property prices.

## 1.1) Business Problem

<img src="./images/vecteezy_debt-vector-icon_18897310.jpg" alt="Housing" height="100">

Home buyers in King County need advice on factors influencing housing prices.

## 1.2) Objectives

We will use data analytics and predictive modeling techniques to provide homeowners, buyers, and investors with valuable insights for making informed decisions in this thriving real estate market.

Specifically, we will:

* Analyze the relationship between house prices and factors such as square footage, number of bedrooms and bathrooms, location, waterfront and view, condition and renovation status.
* Identify the most important factors that affect house prices.
* Develop predictive models that can be used to estimate the future price of a house.
* Share our findings through presentation.
We believe that this project will provide valuable insights into the King County housing market and help people make informed decisions about their real estate investments.

## 1.3) Metric of Success

The project will be considered successful if:
* We accurately make predictions about housing prices and explain factors influencing house prices.
* Our models generate new useful data
* We generate useful recommendations to home buyers to make informed decisions

## 1.4) Understanding Data


The data for this project will be obtained from the King County property appraiser website.
The dataset has the following variables:
* `id` - Unique identifier for a house
* `date` - Date house was sold
* `price` - Sale price (prediction target)
* `bedrooms` - Number of bedrooms
* `bathrooms` - Number of bathrooms
* `sqft_living` - Square footage of living space in the home
* `sqft_lot` - Square footage of the lot
* `floors` - Number of floors (levels) in house
* `waterfront` - Whether the house is on a waterfront
  * Includes Duwamish, Elliott Bay, Puget Sound, Lake Union, Ship Canal, Lake Washington, Lake Sammamish, other lake, and river/slough waterfronts
* `view` - Quality of view from house
  * Includes views of Mt. Rainier, Olympics, Cascades, Territorial, Seattle Skyline, Puget Sound, Lake Washington, Lake Sammamish, small lake / river / creek, and other
* `condition` - How good the overall condition of the house is. Related to maintenance of house.
  * See the [King County Assessor Website](https://info.kingcounty.gov/assessor/esales/Glossary.aspx?type=r) for further explanation of each condition code
* `grade` - Overall grade of the house. Related to the construction and design of the house.
  * See the [King County Assessor Website](https://info.kingcounty.gov/assessor/esales/Glossary.aspx?type=r) for further explanation of each building grade code
* `sqft_above` - Square footage of house apart from basement
* `sqft_basement` - Square footage of the basement
* `yr_built` - Year when house was built
* `yr_renovated` - Year when house was renovated
* `zipcode` - ZIP Code used by the United States Postal Service
* `lat` - Latitude coordinate
* `long` - Longitude coordinate
* `sqft_living15` - The square footage of interior housing living space for the nearest 15 neighbors
* `sqft_lot15` - The square footage of the land lots of the nearest 15 neighbors

## 1.5) Data Relevance and Validation

The data available is relevant for the intended analysis.

## 2.0) Reading and Understanding the Data

In [1]:
# importing necessary libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib 
from matplotlib import pyplot as plt
import plotly as px
from plotly.subplots import make_subplots
import matplotlib.pyplot as plt
import plotly.graph_objects as go
import plotly.express as px
from scipy import stats

In [2]:
# reading the data
data = pd.read_csv('./data/kc_house_data.csv')
# setting the maximum number of columns that will be displayed
pd.set_option('display.max_columns', 21)
data.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors waterfront  view  condition          grade  sqft_above  \
0      5650     1.0        NaN  NONE    Average      7 Average        1180   
1      7242     2.0         NO  NONE    Average      7 Average        2170   
2     10000     1.0         NO  NONE    Average  6 Low Average         770   
3      5000     1.0         NO  NONE  Very Good      7 Average        1050   
4      8080     1.0         NO  NONE    Average         8 Good        1680   

  sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0           0.0      1955           0.0    98178  47.5112 -122.257   
1         400.0      1951        1991.0    98125  47.7210 -122.319   
2           0.0      1933           NaN    98028  47.7379 -122.233   
3         910.0      1965           0.0    98136  47.5208 -122.393   
4           0.0      1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503

In [3]:
data.tail()

id        date     price  bedrooms  bathrooms  sqft_living  \
21592   263000018   5/21/2014  360000.0         3       2.50         1530   
21593  6600060120   2/23/2015  400000.0         4       2.50         2310   
21594  1523300141   6/23/2014  402101.0         2       0.75         1020   
21595   291310100   1/16/2015  400000.0         3       2.50         1600   
21596  1523300157  10/15/2014  325000.0         2       0.75         1020   

       sqft_lot  floors waterfront  view condition      grade  sqft_above  \
21592      1131     3.0         NO  NONE   Average     8 Good        1530   
21593      5813     2.0         NO  NONE   Average     8 Good        2310   
21594      1350     2.0         NO  NONE   Average  7 Average        1020   
21595      2388     2.0        NaN  NONE   Average     8 Good        1600   
21596      1076     2.0         NO  NONE   Average  7 Average        1020   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
21592           0.0      2009           0.0    98103  47.6993 -122.346   
21593           0.0      2014           0.0    98146  47.5107 -122.362   
21594           0.0      2009           0.0    98144  47.5944 -122.299   
21595           0.0      2004           0.0    98027  47.5345 -122.069   
21596           0.0      2008           0.0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
21592           1530        1509  
21593           1830        7200  
21594           1020        2007  
21595           1410        1287  
21596           1020        1357

In [4]:
data.shape

(21597, 21)

In [5]:
data.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  object 
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

In [7]:
data.describe(include = 'all')

id       date         price      bedrooms     bathrooms  \
count   2.159700e+04      21597  2.159700e+04  21597.000000  21597.000000   
unique           NaN        372           NaN           NaN           NaN   
top              NaN  6/23/2014           NaN           NaN           NaN   
freq             NaN        142           NaN           NaN           NaN   
mean    4.580474e+09        NaN  5.402966e+05      3.373200      2.115826   
std     2.876736e+09        NaN  3.673681e+05      0.926299      0.768984   
min     1.000102e+06        NaN  7.800000e+04      1.000000      0.500000   
25%     2.123049e+09        NaN  3.220000e+05      3.000000      1.750000   
50%     3.904930e+09        NaN  4.500000e+05      3.000000      2.250000   
75%     7.308900e+09        NaN  6.450000e+05      4.000000      2.500000   
max     9.900000e+09        NaN  7.700000e+06     33.000000      8.000000   

         sqft_living      sqft_lot        floors waterfront   view condition  \
count   21597.000000  2.159700e+04  21597.000000      19221  21534     21597   
unique           NaN           NaN           NaN          2      5         5   
top              NaN           NaN           NaN         NO   NONE   Average   
freq             NaN           NaN           NaN      19075  19422     14020   
mean     2080.321850  1.509941e+04      1.494096        NaN    NaN       NaN   
std       918.106125  4.141264e+04      0.539683        NaN    NaN       NaN   
min       370.000000  5.200000e+02      1.000000        NaN    NaN       NaN   
25%      1430.000000  5.040000e+03      1.000000        NaN    NaN       NaN   
50%      1910.000000  7.618000e+03      1.500000        NaN    NaN       NaN   
75%      2550.000000  1.068500e+04      2.000000        NaN    NaN       NaN   
max     13540.000000  1.651359e+06      3.500000        NaN    NaN       NaN   

            grade    sqft_above sqft_basement      yr_built  yr_renovated  \
count       21597  21597.000000         21597  21597.000000  17755.000000   
unique         11           NaN           304           NaN           NaN   
top     7 Average           NaN           0.0           NaN           NaN   
freq         8974           NaN         12826           NaN           NaN   
mean          NaN   1788.596842           NaN   1970.999676     83.636778   
std           NaN    827.759761           NaN     29.375234    399.946414   
min           NaN    370.000000           NaN   1900.000000      0.000000   
25%           NaN   1190.000000           NaN   1951.000000      0.000000   
50%           NaN   1560.000000           NaN   1975.000000      0.000000   
75%           NaN   2210.000000           NaN   1997.000000      0.000000   
max           NaN   9410.000000           NaN   2015.000000   2015.000000   

             zipcode           lat          long  sqft_living15     sqft_lot15  
count   21597.000000  21597.000000  21597.000000   21597.000000   21597.000000  
unique           NaN           NaN           NaN            NaN            NaN  
top              NaN           NaN           NaN            NaN            NaN  
freq             NaN           NaN           NaN            NaN            NaN  
mean    98077.951845     47.560093   -122.213982    1986.620318   12758.283512  
std        53.513072      0.138552      0.140724     685.230472   27274.441950  
min     98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%     98033.000000     47.471100   -122.328000    1490.000000    5100.000000  
50%     98065.000000     47.571800   -122.231000    1840.000000    7620.000000  
75%     98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max     98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [8]:
data.isnull().sum()

id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2376
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

## 2.1) Data Cleaning

Observation: as seen in the DataFrame info above sqft_basement is formated as a string instead of float because it contains unwated character

In [9]:
# removing unwanted characters
data["sqft_basement"].replace(r"\?", 0, inplace=True, regex=True)

# casting to float
data['sqft_basement']=data['sqft_basement'].astype(float)

# confirming data type
data['sqft_basement'].dtype

dtype('float64')

In [10]:
#Checking for the year built and year renovated column to understand the data
print(data.yr_built.unique())

print(data.yr_renovated.unique())

[1955 1951 1933 1965 1987 2001 1995 1963 1960 2003 1942 1927 1977 1900
 1979 1994 1916 1921 1969 1947 1968 1985 1941 1915 1909 1948 2005 1929
 1981 1930 1904 1996 2000 1984 2014 1922 1959 1966 1953 1950 2008 1991
 1954 1973 1925 1989 1972 1986 1956 2002 1992 1964 1952 1961 2006 1988
 1962 1939 1946 1967 1975 1980 1910 1983 1978 1905 1971 2010 1945 1924
 1990 1914 1926 2004 1923 2007 1976 1949 1999 1901 1993 1920 1997 1943
 1957 1940 1918 1928 1974 1911 1936 1937 1982 1908 1931 1998 1913 2013
 1907 1958 2012 1912 2011 1917 1932 1944 1902 2009 1903 1970 2015 1934
 1938 1919 1906 1935]
[   0. 1991.   nan 2002. 2010. 1992. 2013. 1994. 1978. 2005. 2003. 1984.
 1954. 2014. 2011. 1983. 1945. 1990. 1988. 1977. 1981. 1995. 2000. 1999.
 1998. 1970. 1989. 2004. 1986. 2007. 1987. 2006. 1985. 2001. 1980. 1971.
 1979. 1997. 1950. 1969. 1948. 2009. 2015. 1974. 2008. 1968. 2012. 1963.
 1951. 1962. 1953. 1993. 1996. 1955. 1982. 1956. 1940. 1976. 1946. 1975.
 1964. 1973. 1957. 1959. 1960. 1967. 1965. 19

Observation: yr_renovated variable is formated as a float and has missing values

In [11]:
# Fill null values, remove decimal point and trailing zeros
data['yr_renovated'] = data['yr_renovated'].fillna(0).astype(int)

# Print unique values
print(data['yr_renovated'].unique())

[   0 1991 2002 2010 1992 2013 1994 1978 2005 2003 1984 1954 2014 2011
 1983 1945 1990 1988 1977 1981 1995 2000 1999 1998 1970 1989 2004 1986
 2007 1987 2006 1985 2001 1980 1971 1979 1997 1950 1969 1948 2009 2015
 1974 2008 1968 2012 1963 1951 1962 1953 1993 1996 1955 1982 1956 1940
 1976 1946 1975 1964 1973 1957 1959 1960 1967 1965 1934 1972 1944 1958]


In [12]:
data.isnull().sum()

id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2376
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated        0
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

In [13]:
data.waterfront.value_counts()

NO     19075
YES      146
Name: waterfront, dtype: int64

Observation: waterfront has missing values

In [14]:
# To preserve the data distribution fill missing values in the 'category_column' with the most frequent category
data['waterfront'].fillna(data['waterfront'].mode()[0], inplace=True)

In [15]:
data.view.value_counts()

NONE         19422
AVERAGE        957
GOOD           508
FAIR           330
EXCELLENT      317
Name: view, dtype: int64

Observation: View has missing values

In [16]:
# Fill missing values in the 'view' with the most frequent category
data['view'].fillna(data['view'].mode()[0], inplace=True)

In [17]:
data.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

Observation: There are no missing values

## 2.2) Analysis

In [18]:
# understanding the influence of waterfront properties on price.

fig = px.histogram(data, x='price', color="waterfront",
                   barmode='group', height=700, width=2000,
                   title='Influence of Waterfront Properties on Price',
                   nbins=50)  # Adjust the nbins parameter as needed
fig.show()

Observation: There are few waterfront properties across many price ranges

In [19]:
#  understanding the influence of properties condition on price.

# Calculating the counts and price range for each condition
condition_summary = data.groupby('condition')['price'].agg(['count', 'min', 'max']).reset_index()

# Sorting condition_summary by 'count' in descending order
condition_summary = condition_summary.sort_values(by='count', ascending=False)

# Creating a bar plot with color representing condition
fig = px.bar(condition_summary, x='condition', y='count', color='condition',
             title='Influence of Property Condition on Price',
             labels={'condition': 'Condition', 'count': 'Count'},
             height=700, width=1500,
             color_discrete_sequence=px.colors.qualitative.Set1)

# Adding text annotations for price range within each bar
for index, row in condition_summary.iterrows():
    price_range_text = f"Price Range: ${row['min']}-{row['max']}"
    fig.add_annotation(
        x=row['condition'], y=row['count'], text=price_range_text,
        showarrow=False, font=dict(size=12)
    )

# Showing the plot
fig.show()


Observation: Houses in Good condition attract the best price range

In [20]:
# understanding property grades against price

# Calculating the count of each grade
grade_counts = data['grade'].value_counts().sort_index().reset_index()
grade_counts.columns = ['grade', 'count']

# Creating a bar plot
fig = px.bar(grade_counts, x='grade', y='count',
             title='Distribution of Property Grades',
             labels={'grade': 'Grade', 'count': 'Count'},
             height=700, width=1000)

# Showing the plot
fig.show()

In [21]:

# Grouping data by grade
grade_groups = data.groupby('grade').mean().sort_index().reset_index()


# Creating a bar plot
fig = px.bar(grade_groups,
             x='grade', y='price',
             title='Mean Sale Price by Grade',
             labels={'grade': 'Grade', 'price': 'Price'},
             height=700, width=1000)

# Showing the plot
fig.show()

Observation: grade 13 Mansion has the highest average price but they were the fewest in number.

In [23]:
# understanding influence of lot area on price

# Creating a scatterplot
fig = px.scatter(data, x='sqft_lot', y='price',
             title='Influence of Property Lot Area on Prices',
             labels={'sqft_lot': 'Lot Area', 'price': 'Price'},
             height=700, width=1000)

# Showing the plot
fig.show()

Observation: Lot Area does not seem to have a linear relationship with price

In [24]:
# checking for outliers in Floors, Bedrooms and Bathrooms Variables

fig = go.Figure()

fig.add_trace(
    go.Box(y= data['floors'], name="Floors"),
)
fig.add_trace(
    go.Box(y= data['bedrooms'], name="Bedrooms"),
)

fig.add_trace(
    go.Box(y= data['bathrooms'], name="bathrooms"),
)
fig.update_layout(
    title_text="Checking for Outliers",
    yaxis_title_text="Values",
    height=1000,
    width=1000
)
fig.show()

In [25]:
# checking for outliers in price Variable
fig = go.Figure()

fig.add_trace(
    go.Box(y= data['price'], name="Price"),
)


fig.update_layout(
    title_text="Checking for Outliers",
    yaxis_title_text="Values",
    height=1000,
    width=1000
)
fig.show()

Observation: The variables have outliers but for our model we will use the data as is without correcting for outliers to maintain robustness of the model and maintain distribution characteristics.

In [26]:
# Calculating the correlation matrix
correlation_matrix = data.corr()

# Creating a heatmap
heatmap = go.Heatmap(z=correlation_matrix.values,
                     x=correlation_matrix.index,
                     y=correlation_matrix.columns,
                     colorscale='Viridis',
                     colorbar=dict(title='Correlation'))

# Extracting correlation values for annotations
annotations = []
for i, row in enumerate(correlation_matrix.values):
    for j, value in enumerate(row):
        annotations.append(
            go.layout.Annotation(
                text="{:.2f}".format(value),
                x=correlation_matrix.index[j],
                y=correlation_matrix.columns[i],
                xref="x1",
                yref="y1",
                showarrow=False,
            )
        )

layout = go.Layout(
    title='Correlation of Numerical Variables',
    xaxis=dict(title='Variables'),
    yaxis=dict(title='Variables'),
    xaxis1=dict(tickangle=-45),
    annotations=annotations
)

fig = go.Figure(data=[heatmap], layout=layout)

# Showing the heatmap
fig.show()


Observation: most of our independent variables have relatively low multicollinearity. However, others like number of bathrooms and square footage of living area have high correlation as expected since having more bathrooms in a property means the property is bigger. Due to lack of control of such we decided to keep all the variables in the model.

## 2.3) Building Linear Regression Model

In [27]:
# Testing for normality

# Selecting the columns for normality testing
columns_to_test = ['price', 'bathrooms', 'sqft_living', 'sqft_above', 'sqft_living15']

# Seting the significance level
alpha = 0.05

# Performing the Anderson-Darling test for normality for each column
for column in columns_to_test:
    selected_data = data[column]
    result = stats.anderson(selected_data, dist='norm')

    # Checking if the data follows a normal distribution based on the test statistic
    if result.statistic > result.critical_values[2]:  
        print(f'The {column} data is not normally distributed (statistic = {result.statistic})')
    else:
        print(f'The {column} data may be normally distributed (statistic = {result.statistic})')

The price data is not normally distributed (statistic = 1193.6731507843288)
The bathrooms data is not normally distributed (statistic = 387.59101464961714)
The sqft_living data is not normally distributed (statistic = 327.3007598772747)
The sqft_above data is not normally distributed (statistic = 589.0920120152878)
The sqft_living15 data is not normally distributed (statistic = 366.65630897079245)


In [28]:
# Checking numeric variables
# Selecting all numeric columns (int and float)
numeric_columns = data.select_dtypes(include=['int64', 'float64'])

# Getting the list of numeric column names
numeric_column_names = numeric_columns.columns.tolist()
numeric_column_names

['id',
 'price',
 'bedrooms',
 'bathrooms',
 'sqft_living',
 'sqft_lot',
 'floors',
 'sqft_above',
 'sqft_basement',
 'yr_built',
 'zipcode',
 'lat',
 'long',
 'sqft_living15',
 'sqft_lot15']

In [29]:
# testing for linear relationship
# Set the significance level
alpha = 0.05

# Looping through each numeric column
for numeric_column in numeric_column_names:
    # Performing Pearson's correlation test
    corr_coefficient, p_value = stats.pearsonr(data[numeric_column], data['price'])

    # Checking if there is a linear relationship
    if p_value < alpha:
        print(f'There is evidence of a linear relationship between {numeric_column} and price (p-value = {p_value})')
    else:
        print(f'There is no evidence of a linear relationship between {numeric_column} and price (p-value = {p_value})')


There is evidence of a linear relationship between id and price (p-value = 0.0137086830163108)
There is evidence of a linear relationship between price and price (p-value = 0.0)
There is evidence of a linear relationship between bedrooms and price (p-value = 0.0)
There is evidence of a linear relationship between bathrooms and price (p-value = 0.0)
There is evidence of a linear relationship between sqft_living and price (p-value = 0.0)
There is evidence of a linear relationship between sqft_lot and price (p-value = 5.571622021561519e-40)
There is evidence of a linear relationship between floors and price (p-value = 2.57e-322)
There is evidence of a linear relationship between sqft_above and price (p-value = 0.0)
There is evidence of a linear relationship between sqft_basement and price (p-value = 0.0)
There is evidence of a linear relationship between yr_built and price (p-value = 2.1175812951495855e-15)
There is evidence of a linear relationship between zipcode and price (p-value = 4.

In [30]:
# a function to get all columns of object type
obj_list = data.select_dtypes(include = "object").columns
print (obj_list)

Index(['date', 'waterfront', 'view', 'condition', 'grade'], dtype='object')


In [31]:
# Label encoding for object to numeric conversion
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

for feat in obj_list:
    data[feat] = le.fit_transform(data[feat])

print (data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  int32  
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     21597 non-null  int32  
 9   view           21597 non-null  int32  
 10  condition      21597 non-null  int32  
 11  grade          21597 non-null  int32  
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  float64
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   21597 non-null  int32  
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

In [32]:
# Dividing the data into features and target variable
X = data.drop(['price', 'date', 'id'], axis=1)
y = data['price'] 


In [40]:
predictors = list(X.columns)
predictors

['bedrooms',
 'bathrooms',
 'sqft_living',
 'sqft_lot',
 'floors',
 'waterfront',
 'view',
 'condition',
 'grade',
 'sqft_above',
 'sqft_basement',
 'yr_built',
 'yr_renovated',
 'zipcode',
 'lat',
 'long',
 'sqft_living15',
 'sqft_lot15']

In [41]:
# Modelling using statsmodels

from statsmodels.formula.api import ols
multiple_formula = 'price ~ bedrooms + bathrooms + sqft_living + sqft_lot + floors + \
 waterfront + condition + grade + sqft_above + sqft_basement + yr_built + yr_renovated + \
 zipcode + lat + long + sqft_living15 + sqft_lot15'
multiple_model = ols(multiple_formula, data).fit()
multiple_summary = multiple_model.summary()
multiple_summary

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.673
Model:                            OLS   Adj. R-squared:                  0.672
Method:                 Least Squares   F-statistic:                     2609.
Date:                Mon, 11 Sep 2023   Prob (F-statistic):               0.00
Time:                        02:51:38   Log-Likelihood:            -2.9533e+05
No. Observations:               21597   AIC:                         5.907e+05
Df Residuals:                   21579   BIC:                         5.908e+05
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept     -1.152e+07   3.03e+06     -3.796      0.000   -1.75e+07   -5.57e+06
bedrooms      -4.387e+04   1979.992    -22.156      0.000   -4.78e+04      -4e+04
bathrooms      5.647e+04   3396.577     16.626      0.000    4.98e+04    6.31e+04
sqft_living     157.3553     18.845      8.350      0.000     120.417     194.293
sqft_lot          0.1897      0.050      3.792      0.000       0.092       0.288
floors         3.404e+04   3758.425      9.058      0.000    2.67e+04    4.14e+04
waterfront     7.665e+05   1.77e+04     43.208      0.000    7.32e+05    8.01e+05
condition      1.638e+04   1281.610     12.784      0.000    1.39e+04    1.89e+04
grade         -1.932e+04    677.246    -28.528      0.000   -2.06e+04    -1.8e+04
sqft_above       64.8373     18.865      3.437      0.001      27.860     101.815
sqft_basement    42.8122     18.706      2.289      0.022       6.147      79.477
yr_built      -1856.1042     74.133    -25.038      0.000   -2001.410   -1710.798
yr_renovated     39.3297      4.164      9.445      0.000      31.168      47.492
zipcode        -567.2305     34.325    -16.525      0.000    -634.510    -499.951
lat             6.57e+05    1.1e+04     59.548      0.000    6.35e+05    6.79e+05
long           -3.24e+05   1.36e+04    -23.908      0.000   -3.51e+05   -2.97e+05
sqft_living15    79.2854      3.393     23.368      0.000      72.635      85.936
sqft_lot15       -0.4369      0.077     -5.703      0.000      -0.587      -0.287
==============================================================================
Omnibus:                    16298.214   Durbin-Watson:                   1.993
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          1142434.214
Skew:                           3.038   Prob(JB):                         0.00
Kurtosis:                      38.109   Cond. No.                     2.13e+08
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.13e+08. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [33]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [43]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# Creating a Linear Regression model
model = LinearRegression()

# Fitting the model to the training data
model.fit(X_train, y_train)

# intercept
print('intercept',model.intercept_)

# Coefficients for all variables
coefficients = model.coef_
variable_names = list(X.columns)
for i, coef in enumerate(coefficients):
    print(f"Coefficient for {variable_names[i]}: {coef}")

# Predicting on the test data
y_pred = model.predict(X_test)

# Evaluating the model
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse) 
r2 = r2_score(y_test, y_pred)

# Calculating the F-statistic and p-value
from scipy import stats

# Calculating the residual sum of squares (RSS)
rss = np.sum((y_test - y_pred) ** 2)

# Degrees of freedom
n = len(y)
k = len(variable_names)
dof_reg = k - 1
dof_resid = n - k

# Calculate the mean squared error for regression and residuals
mse_reg = rss / dof_reg
mse_resid = rss / dof_resid

# Calculate the F-statistic
f_statistic = mse_reg / mse_resid
p_value = 1 - stats.f.cdf(f_statistic, dof_reg, dof_resid)


# Printing evaluation metrics
print(f"Mean Squared Error (MSE): {mse}")
print(f"Root Mean Squared Error (RMSE): {rmse}")
print(f"R-squared (R2): {r2}")
print(f"F-statistic: {f_statistic}")
print(f"P-value: {p_value}")


intercept -6340563.972160146
Coefficient for bedrooms: -44814.925573082975
Coefficient for bathrooms: 55378.911932557334
Coefficient for sqft_living: 141.01800656747443
Coefficient for sqft_lot: 0.14583862950126067
Coefficient for floors: 35157.664833728624
Coefficient for waterfront: 686388.9088227301
Coefficient for view: -32554.661057758272
Coefficient for condition: 17010.52352679578
Coefficient for grade: -19444.61239366132
Coefficient for sqft_above: 80.36728154854919
Coefficient for sqft_basement: 50.597946361994055
Coefficient for yr_built: -1815.4225049187717
Coefficient for yr_renovated: 37.9443130354845
Coefficient for zipcode: -595.7936853375476
Coefficient for lat: 659864.4644823401
Coefficient for long: -303984.6489353096
Coefficient for sqft_living15: 73.24620750372713
Coefficient for sqft_lot15: -0.5226343925391608
Mean Squared Error (MSE): 42230927811.607
Root Mean Squared Error (RMSE): 205501.64917004196
R-squared (R2): 0.6756862676861248
F-statistic: 1269.35294117647

## 2.4) Conclusion

* The overall model fit is reasonable, with an R-squared value of approximately 0.676. This means that around 67.6% of the variability in house prices can be explained by the combination of independent variables included in the model. 
* Several independent variables show statistically significant relationships with house prices. These include:

 * Bedrooms: The number of bedrooms has a negative effect on house prices.
 * Bathrooms: More bathrooms tend to increase house prices.
 * Square Footage: An increase in square footage positively impacts house prices.
 * Waterfront Property: Having a waterfront location significantly increases house prices.
 * Condition: Better property condition is associated with higher house prices.
 * Grade: Higher property grades positively affect house prices.
 * Year Built: Older homes tend to have lower prices, while newer homes have higher prices.
 * Year Renovated: Renovations positively impact house prices.
 * Location (Zipcode, Latitude, Longitude): These location-related variables significantly influence house prices.

## 2.5) Recommendation

* Property Improvement: Consider investing in property improvements, such as renovating or upgrading certain features, to potentially increase the property's value.
* Location Matters: The location of the property, as indicated by variables like zipcode, latitude, and longitude, plays a significant role in house prices.
* Bedrooms and Bathrooms: The number of bedrooms and bathrooms can affect prices. Sellers should highlight these features, and buyers should consider their needs.
* Square Footage: Increasing square footage can positively impact property value. Sellers may benefit from maximizing usable living space.
* Waterfront Properties: Waterfront properties command higher prices. Buyers interested in such properties should expect premium pricing.
* Age of the Property: Older properties tend to have lower prices, so buyers should consider the trade-offs between historic charm and modern amenities.
* Condition and Grade: Sellers should focus on property condition and grade to potentially increase market appeal and pricing.

In summary, this regression analysis provides insights into the factors influencing house prices in the dataset. However, it's essential to consider these results as a tool to inform decisions, rather than a definitive predictor. Real estate prices can be influenced by various external factors, and local market conditions may vary. Continual analysis and market awareness are key to making informed decisions in the real estate market.